### PR TITLE
Add `MasgnNode` class for `masgn` nodes

### PR DESCRIPTION
**Pull request by dvandersluis**
Follows #201. I actually ended up needing `masgn` to have its own class with `expression` defined in order to do the refactoring I wanted to, so this adds it.

Because `masgn` nodes are more involved than the other assignment types, this class has a different interface than the other assignments, except for `expression` which works the same. Rather than `name` that the other assignment classes have, `MasgnNode` has `names` (since there will inherently be more than one). It also is able to handle nested `mlhs`es.

### CODE REVIEW

**Comment by dvandersluis** [comment]
Sorry I missed this in my previous PR @marcandre 😅 🙌

**Review by marcandre — CHANGES_REQUESTED** [review]
`masgn` is complicated, isn't it? 😅

PR looks like a good start, but you seem to have overlooked 2 important cases: `foo.bar, foo[:baz], ... = ...`.
Please describe results of your proposed methods and add tests.

Also:
* `each_descendant` should only be used on the first child. Please add a test for `x, y = 1, z+=2` for example. Even then, some corner cases could be problematic: `foo[x+=1], bar = ...`

* I doubt `each_assignment` is necessary. Performance of enumerators is typically quite poor, especially for short lists

**Comment by dvandersluis** [comment]
I knew there were things I wasn't thinking of, thanks @marcandre I'll fix up for those cases!

**Comment by dvandersluis** [comment]
@marcandre updated now.

I have now introduced a `MlhsNode` as well, which the `assignments` logic has been moved to (and `each_assignment` has been removed). `MlhsNode#assignments` takes care of finding nested assignments (in splats or other `mlhs` nodes), and also now returns `send` nodes to cover your edge cases. I also have a tests covering `casgn` nodes now, which were previously missing.

`MasgnNode#names` is now able to handle `send` nodes as well (by just returning the source, I don't think there's a better solution for `foo.bar` or `foo[:bar]`, etc.).

For symmetry, `MasgnNode` now has `lhs` (returns the `mlhs` node) and `rhs` (alias of `expression`) methods.

Please let me know your thoughts!

**Comment by dvandersluis** [comment]
The "modern" spec was failing on `indexasgn` nodes so that's handled now too.

File: lib/rubocop/ast/node/masgn_node.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      end
+      alias rhs expression
+
+      # @return [Array<Node>] values being assigned on the RHS of the multiple assignment
+      def values
+        array? ? expression.children : [expression]
+      end
+
+      # @return [Boolean] whether the expression has multiple values
+      def array?
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply lib/rubocop/ast/node/masgn_node.rb:45]: I can't remember, so I'll remove it.

File: lib/rubocop/ast/node/masgn_node.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      # @return [Node] the expression being assigned.
+      def expression
+        node_parts[1]
+      end
+      alias rhs expression
+
+      # @return [Array<Node>] values being assigned on the RHS of the multiple assignment
+      def values
+        array? ? expression.children : [expression]
+      end
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply lib/rubocop/ast/node/masgn_node.rb:42]: To clarify, are you suggestion that the method is not useful? My thinking was there'd be use being able to collect the RHS value nodes for processing. I don't have any specific context anymore due to the time gap but I recall working with multiple assign is complicated.

File: spec/rubocop/ast/masgn_node_spec.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    context 'with nested `mlhs` nodes' do
+      let(:source) { 'a, (b, c) = z' }
+
+      it { is_expected.to eq(%i[a b c]) }
+    end
+
+    context 'with nested assignment on LHS' do
+      let(:source) { 'a, b[c+=1] = z' }
+
+      it { is_expected.to eq([:a, 'b[c+=1]']) }
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb:29]: Changed to `:"b[]"`

File: spec/rubocop/ast/masgn_node_spec.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    end
+
+    context 'with a method chain on LHS' do
+      let(:source) { 'a, b.c = z' }
+
+      it { is_expected.to eq([:a, 'b.c']) }
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb:35]: Changed to `:c` (for consistency with above, the `=` is left out), does that work for you?

File: spec/rubocop/ast/masgn_node_spec.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      let(:source) { 'x, y = *z' }
+
+      it { is_expected.to eq(s(:array, s(:splat, s(:send, nil, :z)))) }
+    end
+
+    context 'with assignment on RHS' do
+      let(:source) { 'x, y = 1, z+=2' }
+
+      it { is_expected.to eq(s(:array, s(:int, 1), s(:op_asgn, s(:lvasgn, :z), :+, s(:int, 2)))) }
+    end
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb:70]: I thought it was useful to have examples of potential uses, what single test would you have in mind?

File: lib/rubocop/ast/node/masgn_node.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (dvandersluis) [reply lib/rubocop/ast/node/masgn_node.rb:42]: Updated so that `x, y = z` and `x, y = [z]` do not produce the same values, added test.

File: spec/rubocop/ast/masgn_node_spec.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    include AST::Sexp
+
+    subject { masgn_node.values }
+
+    context 'when the RHS has a single value' do
+      let(:source) { 'x, y = z' }
+
+      it { is_expected.to eq([s(:send, nil, :z)]) }
+    end
+
```
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb:83]: Done.

**Comment by dvandersluis** [comment]
@marcandre it's been a long time but I updated this PR and tried to address your concerns/simplify some things.

File: spec/rubocop/ast/masgn_node_spec.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    context 'with nested `mlhs` nodes' do
+      let(:source) { 'a, (b, c) = z' }
+
+      it { is_expected.to eq(%i[a b c]) }
+    end
+
+    context 'with assignment inside `send` node on LHS' do
+      let(:source) { 'a, b[c+=1] = z' }
+
+      it { is_expected.to eq([:a, :'b[]']) } # rubocop:disable Style/SymbolArray
```
Comment (dvandersluis) [inline spec/rubocop/ast/masgn_node_spec.rb:29]: Explicitly disabled `Style/SymbolArray` here because I thought this was better than `%i[a, b\[\]]`

File: spec/rubocop/ast/masgn_node_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcandre) [reply spec/rubocop/ast/masgn_node_spec.rb:70]: IMO, tests should require a reason for existing. If there's no reasonable implementation of a method that would have one test fail and not the other, than the two tests are redundant. I believe this is the case here.

I'd pick the simplest one.

For me

File: spec/rubocop/ast/masgn_node_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcandre) [reply spec/rubocop/ast/masgn_node_spec.rb]: Mmm, thanks, but actually the idea is to _not_ be consistent, as we don't want the same result as the completely different source `a, c = z`.

So setting a variable `a` => `:a`. Calling the `a=` method of some expression => ? (but must be different from `:a`)

File: spec/rubocop/ast/masgn_node_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcandre) [reply spec/rubocop/ast/masgn_node_spec.rb]: Actually, I think I was wrong to suggest `:"b[]"`, as `b` could be abitrarily complicated, as it doesn't have to be a variable name, right?
So maybe it should be `:[]=` or `:[]` instead?

**Comment by marcandre** [comment]
That's great @dvandersluis , good work.

So a decision remains as to what we return for the LHS in the different cases:
* setting a variable `a` => `:a` (probably)
* calling `[]=` => ?
* calling `a=` => ?

File: spec/rubocop/ast/masgn_node_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb:70]: Fair enough, I prefer test to be expressive (clearly 😅) but I'll reduced as requested.

**Comment by dvandersluis** [comment]
I think for simplicity with `names`, if a LHS value is a `send` node, using `method_name` unadulterated would suffice given your suggestions. 

`a, b = z` => `[:a, :b]`
`a, b.c = z` => `[:a, :c=]`
`a, b[c] = z` => `[:a, :[]=]`

Is this what you had in mind @marcandre?

**Comment by marcandre** [comment]
> I think for simplicity with `names`, if a LHS value is a `send` node, using `method_name` unadulterated would suffice given your suggestions.
> 
> `a, b.c, d[e] = z` => `[:a, :c=, :[]=]`
> 
> Is this what you had in mind @marcandre?

Yes, I agree that's probably best 👍

**Comment by dvandersluis** [comment]
Updated `#names` and `#expression` tests as per review.

File: spec/rubocop/ast/masgn_node_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (dvandersluis) [reply spec/rubocop/ast/masgn_node_spec.rb]: Fixed now.

File: lib/rubocop/ast/node/masgn_node.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+      # The expression being assigned to the variable.
+      #
+      # @return [Node] the expression being assigned.
+      def expression
+        node_parts[1]
+      end
+      alias rhs expression
+
+      # @return [Array<Node>] values being assigned on the RHS of the multiple assignment
```
Comment (marcandre) [inline lib/rubocop/ast/node/masgn_node.rb:39]: I think it would be worth explaining the subtle difference between `expression` and `values`, in particular that `[1, 2]` vs `1, 2` (which `expression` considers the same but `values` doesn't)

**Comment by marcandre** [comment]
So quick! A bit of extra doc and this is good to go. It's 2am here in Barcelona, so I'll look at it tomorrow 😅

**Comment by dvandersluis** [comment]
Thanks as always for your quick reviews @marcandre! I updated with some more documentation on `expression` and `values`; let me know if you want more there.

**Comment by marcandre** [comment]
Very nice, thank you ❤️
